### PR TITLE
Reduce beatmap difficulty processing time

### PR DESCRIPTION
**Pull request by smoogipoo**
Reduces computation time by 25% across 20000 beatmaps, testing the osu! mode only.

Mostly just removing allocations to reduce GC wait time. There are certainly more improvements that can be done - bezier calculations are quite inefficient, buffers could be used to reduce more control point list constructions, and beatmap converters could not use linq (or use compiled linq?) to yield hitobjects, etc...

Other game modes (especially mania) will need similar work, but I'll get to that when the diff calc algorithm is on-par there.

### CODE REVIEW

File: osu.Game/Beatmaps/BeatmapConverter.cs [hunk 1]
```diff
@@ -55,39 +55,41 @@ protected virtual Beatmap<T> ConvertBeatmap(IBeatmap original)
...
-                yield break;
-            }
-
-            var converted = ConvertHitObject(original, beatmap).ToList();
-            ObjectConverted?.Invoke(original, converted);
+            var result = new List<T>(hitObjects.Count);
 
-            // Convert the hit object
-            foreach (var obj in converted)
+            // ReSharper disable once ForCanBeConvertedToForeach
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapConverter.cs]: Unnecessary?

File: osu.Game/Beatmaps/Formats/LegacyDecoder.cs [hunk 2]
```diff
@@ -69,7 +69,9 @@ protected virtual void ParseLine(T output, Section section, string line)
 
         protected string StripComments(string line)
         {
-            var index = line.IndexOf("//", StringComparison.Ordinal);
+            const string comment_prefix = "//";
```
Comment (peppy) [inline osu.Game/Beatmaps/Formats/LegacyDecoder.cs]: not sure what the overhead of the string construction is, but if we're going this far, would `var index = line.AsSpan().IndexOf(new Span<char>(new [] { '/', '/' }));` possibly be an improvement?

File: osu.Game/Beatmaps/ControlPoints/ControlPointInfo.cs [hunk 3]
```diff
@@ -12,6 +12,11 @@ namespace osu.Game.Beatmaps.ControlPoints
     [Serializable]
     public class ControlPointInfo
     {
+        private readonly TimingControlPoint timingSearchPoint = new TimingControlPoint();
```
Comment (peppy) [inline osu.Game/Beatmaps/ControlPoints/ControlPointInfo.cs]: makes these methods non-threadsafe. probably worth adding to xmldoc or reconsidering.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Formats/LegacyDecoder.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game/Beatmaps/Formats/LegacyDecoder.cs]: Can just be moved inlined due to string interning. Const is not really useful.

**Review by peppy — APPROVED** [review]

**Pull request merged**